File: typescript/questions/firstMissingPositive.ts
function firstMissingPositive(nums: number[]): number {
  const nl = nums.length;
  for (let i=0; i<nums.length; ++i) {
    while (nums[i]>0 && nums[i]<=nl && nums[nums[i]-1]!=nums[i] ) {
      [nums[nums[i]-1], nums[i]] = [nums[i], nums[nums[i]-1]];
    }
  }
  for (let i=0; i<nl; ++i) {
    if (nums[i] != i+1) {
      return i+1;
    }
  }
  return nl + 1;    
};

File: rust/questions/first_missing_positive.rs
impl Solution {
    pub fn first_missing_positive(mut nums: Vec<i32>) -> i32 {
        let nl = nums.len();
        for p in 0..nl {
            while nums[p] > 0 && nums[p] <= nl as i32 {
                let np = nums[p] as usize;
                if nums[np - 1] == nums[p] {
                    break;
                }
                nums.swap(p, np - 1);
            }
        }
        for p in 0..nl {
            if nums[p] != p as i32 + 1 {
                return p as i32 + 1;
            }
        }
        nl as i32 + 1
    }
}
